Add Docker build and multi-environment instructions (#512)
AlexsLemonade/OpenScPCA-analysis: docs/software-platforms/docker/docker-images.md

COPY conda-lock.yml conda-lock.yml

# restore from conda-lock.yml file and clean up to reduce image size
RUN conda-lock install -n ${ENV_NAME} \
  && conda clean --all --yes

# Activate conda environment on bash launch
RUN echo "conda activate ${ENV_NAME}" >> ~/.bashrc

# Set entrypoint to bash to activate the environment for any commands
ENTRYPOINT ["bash", "-l", "-c"]
```

### Images with both R and conda environments

If your analysis module requires both R and conda environments, you may have to do just a bit more work to set up the Dockerfile, as there is not a single base image that includes both R and conda environments.

> <!-- Should we make one, maybe? -->

Reviewer: definitely maybe!

We recommend starting with an R-based image and then installing conda manually.

> In the example below, we are adapting the installation steps used in the [official Miniconda Dockerfile](https://github.com/ContinuumIO/docker-images/blob/main/miniconda3/debian/Dockerfile)

Reviewer: ```suggestion
In the example below, we have adapted the installation steps used in the [official Miniconda Dockerfile](https://github.com/ContinuumIO/docker-images/blob/main/miniconda3/debian/Dockerfile).
```

ENV RENV_CONFIG_CACHE_ENABLED FALSE

# Copy conda lock file to image
COPY conda-lock.yml conda-lock.yml

# restore from conda-lock.yml file and clean up to reduce image size
RUN conda-lock install -n ${ENV_NAME} conda-lock.yml && \
  conda clean --all --yes

# Copy the renv.lock file from the host environment to the image
COPY renv.lock renv.lock

# restore from renv.lock file and clean up to reduce image size
RUN Rscript -e 'renv::restore()' \
  && rm -rf ~/.cache/R/renv \
  && rm -rf /tmp/downloaded_packages \
  && rm -rf /tmp/Rtmp*

# Activate conda environment on bash launch
RUN echo "conda activate ${ENV_NAME}" >> ~/.bashrc

# Set entrypoint to bash to activate the environment for any commands
ENTRYPOINT ["bash", "-l", "-c"]
```

!!! tip
    Note that we install all of Miniconda, `conda-lock` and `renv` before installing any module-specific packages.
    This is because these steps are not likely to change often, so we can take advantage of Docker's caching to avoid re-installing them every time we build the image.

## Building Docker images

### Local testing

If you have a `Dockerfile` set up for your analysis module, you can build the Docker image locally to test it.

> To match the infrastructure that the OpenScPCA project uses you should use the `docker buildx` system and the `--platform linux/amd64` flag when building the image.

Reviewer: Will this flag build a _usable_ docker image on all machines though? IIRC, `buildx` indeed makes that happen. Yes?

Author: I'm not sure I understand the question? If docker can build it, docker can run it, since building requires running an image.

Reviewer: This question was brought to you by OpenPBTA flashbacks of trying to run in containers when I first got my M1 computer. The fact that you don't understand it sounds like things have since resolved themselves and we can move on?

The `Dockerfile` should be in the root of the analysis module directory.

The command below will build the Docker image using the `Dockerfile` in the module directory and tag it as `openscpca/your-module:latest`.
You should replace `your-module` with the name of your analysis module.

```bash
# move to the module directory
cd /path/to/your-module
# build the docker image
docker buildx build . -t openscpca/your-module:latest --platform linux/amd64
```

### Automated testing and deployment

> Once a module has a complete Docker image, we can add it to the OpenScPCA automated tests and the [OpenScPCA Docker registry](https://gallery.ecr.aws/openscpca).

Reviewer: I think we didn't discuss testing this as a page in reproducibility but that was probably an oversight? What do you think? 

```suggestion
Once a module has a complete Docker image, we can add it to the OpenScPCA automated tests and the [OpenScPCA Docker registry](https://gallery.ecr.aws/openscpca). <!-- STUB_LINK to testing docs -->
```

> Once a module has a complete Docker image, we can add it to the OpenScPCA automated tests and the [OpenScPCA Docker registry](https://gallery.ecr.aws/openscpca). <!-- STUB_LINK to testing docs -->

Reviewer: noting we are planning for gha docs! It's just that those issues haven't been filed yet since things are still moving around. so, good stub ✔️

These steps will usually be completed by Data Lab staff.

To add a module's Docker image to testing, we first activate the Github Action that builds the Docker image whenever the module's Dockerfile or environment files are updated.

> In the `.github/workflows/` directory, there should be a `docker_your-module.yml` file that was created when the module was initialized.

Reviewer: ```suggestion
In the `.github/workflows/` directory, there should be a `docker_{your-module}.yml` file that was created when the module was initialized.
```

> In the `.github/workflows/` directory, there should be a `docker_{your-module}.yml` file that was created when the module was initialized.

Reviewer 2: Does this now automatically happen? When I created my module it didn't, so just wanted to make sure this is included in the module creation script now?

Author: Yes, we added this in #464 (with updates to the exact files since)

To activate the Github Action, we will uncomment the `on:` block at the top of the file and update the file lists to include any files that should trigger a rebuild, such as files that define the software environment or that are otherwise copied into the Docker image.

> The module name will also be added to the `modules:` list in the `.github/workflows/docker_all-modules.yml` file.

Reviewer: ```suggestion
The module name will also be added to the `modules:` list in the `.github/workflows/docker_all-modules.yml` workflow file.
```
